src/Processors/QueryPlan/Optimizations/filterPushDown.cpp

```diff
@@ -328,6 +328,10 @@ size_t tryPushDownFilter(QueryPlan::Node * parent_node, QueryPlan::Nodes & nodes
             if (table_join.kind() != JoinKind::Inner && table_join.kind() != JoinKind::Cross && table_join.kind() != kind)
```

Comment:
Shouldn't we also forbid `SEMI/ANTI` ? Although, I checked on build from this PR and conditions is not pushed in case of ANTI, but I don't see where it's prevented.

Same for `ASOF INNER` it also can affect correctness I suppose:

```sql
SELECT *
FROM (SELECT * FROM Values('id UInt64, t UInt64', (1,3))) AS t1
ASOF JOIN (SELECT * FROM Values('id UInt64, t UInt64', (1,1),(1,2),(1,3),(1,4),(1,5))) AS t2
ON t1.id = t2.id AND t1.t < t2.t
WHERE t2.t != 4
-- expected to get row (1, 3, 1, 4) from JOIN and empty result from the query
```
is not the same as

```sql
SELECT *
FROM (SELECT * FROM Values('id UInt64, t UInt64', (1,3))) AS t1
ASOF JOIN (SELECT * FROM Values('id UInt64, t UInt64', (1,1),(1,2),(1,3),(1,4),(1,5)) WHERE t != 4) AS t2
ON t1.id = t2.id AND t1.t < t2.t
-- expected to get row (1, 3, 1, 5)
```

and it works incorrectly on current master (but correct with `allow_experimental_analyzer`) 😵‍💫